kratos/modeler/voxel_mesh_generator_modeler.cpp

```diff
@@ -0,0 +1,335 @@
+    mMeshingData.SetNumberOfDivisions(mKeyPlanes[0].size(), mKeyPlanes[1].size(), mKeyPlanes[2].size());
+
+    array_1d<double,3> margins = ZeroVector(3);
+    for(int i = 0 ; i < 3 ; i++) {
+        margins[i] = (mKeyPlanes[i].back() - mKeyPlanes[i].front()) * 1.0e-2;
+    }
+
+    const double margin = *std::max_element(margins.begin(), margins.end());
+    mColors.ExtendBoundingBox(mpInputModelPart->Nodes(), margin);
+}
+
+void VoxelMeshGeneratorModeler::ApplyColoring(Parameters ColoringParameters, int OutsideColor){
+    Timer::Start("MeshColoring");
+
+    VoxelMesherColoringFactory factory("coloring strategy");
+    mColors.SetAllColors(OutsideColor);
+
+    for(auto parameters : ColoringParameters){
+
+        if (!parameters.Has("outside_color")) {
+            parameters.AddInt("outside_color", OutsideColor);
+        }
+
+        VoxelMesherColoring::Pointer p_coloring = factory.Create(*this, parameters);
+
+        p_coloring->ValidateParameters();
+
+        std::string model_part_name = parameters["model_part_name"].GetString();
+        std::string type = parameters["type"].GetString();
+        KRATOS_INFO("Modeler") << "Applying color to " << type << " for " << model_part_name  << " model part" << std::endl;
+
+        p_coloring->Apply();
+    }
+    Timer::Stop("MeshColoring");
+}
+
+void VoxelMeshGeneratorModeler::SetStartIds(ModelPart& rTheVolumeModelPart)
+{
+    ModelPart& r_root_model_part = rTheVolumeModelPart.GetRootModelPart();
+    std::size_t start_node_proposal = r_root_model_part.NodesArray().empty() ? 1 : r_root_model_part.NodesArray().back()->Id() + 1;
+    mStartNodeId = std::max(start_node_proposal, mStartNodeId);
+
+    std::size_t start_element_proposal = r_root_model_part.ElementsArray().empty() ? 1 : r_root_model_part.ElementsArray().back()->Id() + 1;
+    mStartElementId = std::max(start_element_proposal, mStartElementId);
+
+    std::size_t start_condition_proposal = r_root_model_part.ConditionsArray().empty()? 1 :r_root_model_part.ConditionsArray().back()->Id() + 1;
+    mStartConditionId = std::max(start_condition_proposal, mStartConditionId);
+}
+
+void VoxelMeshGeneratorModeler::GenerateEntities(
+    ModelPart& rTheVolumeModelPart,
+    Parameters EntityGeneratorParameters
+    )
+{
+    Timer::Start("EntityGeneration");
+    VoxelMesherEntityGenerationFactory factory("entity generation strategy");
+
+    for(auto parameters : EntityGeneratorParameters){
+
+        auto p_generator = factory.Create(*this, parameters);
+        p_generator->ValidateParameters();
+
+        std::string model_part_name = parameters["model_part_name"].GetString();
+        std::string type = parameters["type"].GetString();
+        KRATOS_INFO("Modeler") << "Generate " << type << " in " << model_part_name  << " model part" << std::endl;
+
+        p_generator->Generate();
+    }
+    Timer::Stop("EntityGeneration");
+}
+
+ModelPart& VoxelMeshGeneratorModeler::CreateAndGetModelPart(std::string const& FullName)
```

Comment:
I am pretty sure the `Model` does the same nowadays
@sunethwarna can you confirm pls?

Reply:
I am using the model capabilities now, but I'm leaving a helper function to check for existence of the new model part (as in that case the model method logs a warning)